compiler/src/yul/abi/functions.rs

```diff
@@ -0,0 +1,205 @@
+use crate::yul::abi::utils::{
+    decode_name,
+    encode_name,
+    head_offsets,
+};
+use crate::yul::operations;
+use fe_semantics::namespace::types::{
+    AbiArraySize,
+    AbiDecodeLocation,
+    AbiEncoding,
+    AbiType,
+    AbiUintSize,
+};
+use yultsur::*;
+
+/// Dynamically creates an encoding function for any set of type parameters.
+pub fn encode<T: AbiEncoding>(types: Vec<T>) -> yul::Statement {
```

Comment:
add inline comments explaining whats going on for both functions in here. also, make sure all public functions in this module have api docs